File: native/android/app/src/main/java/com/smartmaps/ae/core/TransportTimeEngine.kt
package com.smartmaps.ae.core

/**
 * A/E Transport Time Engine — bus/train/tube arrival feeds.
 *
 * Providers register themselves and are polled by AECore's tick; whatever
 * they return is merged into one arrivals board keyed by stop. A provider
 * that is unreachable simply contributes nothing this tick — absence of live
 * data is shown as absence (timetable rows carry live=false), never faked.
 * No fallback ladders: the board is exactly the union of what providers know.
 */
interface TransportFeedProvider {
    val id: String
    val modes: Set<TransportMode>

    /** Stops near [center] within [radiusM] this provider knows about. */
    fun stopsNear(center: GeoPoint, radiusM: Double): List<TransitStop>

    /** Current arrivals for a stop. Empty when the provider has nothing. */
    fun arrivals(stopId: String, nowMs: Long): List<TransitArrival>
}

class TransportTimeEngine {

    private val providers = mutableListOf<TransportFeedProvider>()
    private var board: Map<String, List<TransitArrival>> = emptyMap()
    private var stops: List<TransitStop> = emptyList()
    private var lastRefreshMs = 0L
    private val refreshIntervalMs = 15_000L

    fun register(provider: TransportFeedProvider) {
        if (providers.none { it.id == provider.id }) providers.add(provider)
    }

    fun stops(): List<TransitStop> = stops

    fun arrivalsFor(stopId: String): List<TransitArrival> = board[stopId] ?: emptyList()

    /** All arrivals across nearby stops, soonest first. */
    fun allArrivals(): List<TransitArrival> =
        board.values.flatten().sortedBy { it.expectedAtMs }

    /**
     * Refresh the board around [center]. Rate-limited internally; callers can
     * invoke every tick. Providers that throw are skipped this cycle —
     * their data is simply absent until they recover.
     */
    fun refresh(center: GeoPoint, nowMs: Long, radiusM: Double = 800.0) {
        if (nowMs - lastRefreshMs < refreshIntervalMs && board.isNotEmpty()) return
        lastRefreshMs = nowMs

        val allStops = mutableListOf<TransitStop>()
        val newBoard = mutableMapOf<String, MutableList<TransitArrival>>()
        for (provider in providers) {
            val providerStops = runCatching { provider.stopsNear(center, radiusM) }.getOrDefault(emptyList())
            for (stop in providerStops) {
                if (allStops.none { it.id == stop.id }) allStops.add(stop)
                val arrivals = runCatching { provider.arrivals(stop.id, nowMs) }.getOrDefault(emptyList())
                newBoard.getOrPut(stop.id) { mutableListOf() }.addAll(arrivals)
            }
        }
        newBoard.values.forEach { it.sortBy { a -> a.expectedAtMs } }
        stops = allStops.sortedBy { haversineM(center, it.point) }
        board = newBoard
    }

    fun reset() {
        board = emptyMap()
        stops = emptyList()
        lastRefreshMs = 0
    }
}

File: native/ios/SMCore/Core/TransportTimeEngine.swift
// TransportTimeEngine.swift — A/E Transport Time: bus/train/tube feeds.
//
// Providers register and are polled by AECore's tick; the board is exactly
// the union of what providers know. Absence of live data is absence — never
// faked, no fallback ladders. Mirrors the Kotlin engine.

import Foundation

protocol TransportFeedProvider {
    var id: String { get }
    var modes: Set<TransportMode> { get }
    func stopsNear(center: GeoPoint, radiusM: Double) -> [TransitStop]
    func arrivals(stopId: String, nowMs: Int64) -> [TransitArrival]
}

final class TransportTimeEngine {

    private var providers: [TransportFeedProvider] = []
    private var board: [String: [TransitArrival]] = [:]
    private var nearbyStops: [TransitStop] = []
    private var lastRefreshMs: Int64 = 0
    private let refreshIntervalMs: Int64 = 15_000

    init() {}

    func register(_ provider: TransportFeedProvider) {
        if !providers.contains(where: { $0.id == provider.id }) {
            providers.append(provider)
        }
    }

    func stops() -> [TransitStop] { nearbyStops }

    func arrivalsFor(stopId: String) -> [TransitArrival] { board[stopId] ?? [] }

    func allArrivals() -> [TransitArrival] {
        board.values.flatMap { $0 }.sorted { $0.expectedAtMs < $1.expectedAtMs }
    }

    func refresh(center: GeoPoint, nowMs: Int64, radiusM: Double = 800) {
        if nowMs - lastRefreshMs < refreshIntervalMs && !board.isEmpty { return }
        lastRefreshMs = nowMs

        var allStops: [TransitStop] = []
        var newBoard: [String: [TransitArrival]] = [:]
        for provider in providers {
            let providerStops = provider.stopsNear(center: center, radiusM: radiusM)
            for stop in providerStops {
                if !allStops.contains(where: { $0.id == stop.id }) { allStops.append(stop) }
                let arrivals = provider.arrivals(stopId: stop.id, nowMs: nowMs)
                newBoard[stop.id, default: []].append(contentsOf: arrivals)
            }
        }
        for key in newBoard.keys {
            newBoard[key]?.sort { $0.expectedAtMs < $1.expectedAtMs }
        }
        nearbyStops = allStops.sorted { haversineM(center, $0.point) < haversineM(center, $1.point) }
        board = newBoard
    }

    func reset() {
        board = [:]
        nearbyStops = []
        lastRefreshMs = 0
    }
}
